### ISSUE TITLE
resume fine-tuning from an intermediate checkpoint

### ISSUE THREAD

**Issue by aniketmaurya** [Lightning-AI/lit-llama#180]
It would be awesome to provide the functionality to resume fine-tuning from an intermediate checkpoint.

**Comment by laosuan** [comment]
especially resume finetune adapter

**Comment by chakt** [comment]
Sorry, may I ask if the following code modification will work (modified in bold)? Will it resume from the mid-checkpoint.pth and continue the training correctly?


def main(
    data_dir: str = "data/alpaca", 
    pretrained_path: str = "checkpoints/lit-llama/7B/lit-llama.pth",
    out_dir: str = "out/adapter/alpaca",
    **adapter_path: str = None,**
):

    fabric = L.Fabric(
        accelerator="cuda", 
        devices=devices, 
        strategy=(DeepSpeedStrategy(config=ds_config) if devices > 1 else "auto"), 
        precision="bf16-mixed",
    )
    fabric.launch()
    fabric.seed_everything(1337 + fabric.global_rank)

    if fabric.global_rank == 0:
        os.makedirs(out_dir, exist_ok=True)

    train_data, val_data = load_datasets(data_dir=data_dir)

    config = LLaMAConfig()
    config.block_size = block_size

    if not os.path.isfile(pretrained_path):
        raise FileNotFoundError(
            f"Can't find the pretrained weights at {pretrained_path}."
            " Please follow the instructions in the README to download them."
        )
    checkpoint = torch.load(pretrained_path)

    with fabric.device:
        torch.set_default_tensor_type(torch.HalfTensor)
        model = LLaMA(config).bfloat16()
        torch.set_default_tensor_type(torch.FloatTensor)
        **state_dict = checkpoint**
        **# Load the fine-tuned adapter weights, if provided
        if adapter_path is not None and os.path.isfile(adapter_path):
            adapter_checkpoint = torch.load(adapter_path)
            state_dict.update(adapter_checkpoint)

        model.load_state_dict(state_dict, strict=False)**
    
    mark_only_adapter_as_trainable(model)

    num_params = sum([p.numel() for p in model.parameters() if p.requires_grad])
    print(f"Number of trainable parameters: {num_params}")

    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay)
    model, optimizer = fabric.setup(model, optimizer)
    **train(fabric, model, optimizer, train_data, val_data, out_dir, adapter_path)**

    # Save the final checkpoint at the end of training
    save_model_checkpoint(fabric, model, os.path.join(out_dir, "lit-llama-adapter-finetuned.pth"))


def train(
    fabric: L.Fabric,
    model: torch.nn.Module,
    optimizer: torch.optim.Optimizer,
    train_data: np.ndarray,
    val_data: np.ndarray,
    out_dir: str,
    **adapter_path: str,**
) -> None:
    """The training loop.

    Loosely based on the nanoGPT implementation: https://github.com/karpathy/nanoGPT.
    """
    **import re**

    **# Extract the iteration number from the adapter path
    iter_match = re.search(r"iter-(\d+).pth", adapter_path)
    starting_iter = int(iter_match.group(1)) if iter_match else 0**

    step_count = 0

    **for iter_num in range(starting_iter, starting_iter + max_iters):**

        if step_count <= warmup_steps:
            # linear warmup
            lr = learning_rate * step_count / warmup_steps
            for param_group in optimizer.param_groups:
                param_group['lr'] = lr

        t0 = time.time()

        input_ids, targets = get_batch(fabric, train_data)
        logits = model(input_ids)
        loss = loss_fn(logits, targets)
        with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_steps != 0)):
            fabric.backward(loss / gradient_accumulation_steps)

        if (iter_num + 1) % gradient_accumulation_steps == 0:
            optimizer.step()
            optimizer.zero_grad()
            step_count += 1

            if step_count % eval_interval == 0:
                val_loss = validate(fabric, model, val_data)
                fabric.print(f"step {step_count}: iter {iter_num}: val loss {val_loss:.4f}")
                fabric.barrier()

            if step_count % save_interval == 0:
                print(f"Saving adapter weights to {out_dir}")
                # TODO: Provide a function/script to merge the adapter weights with pretrained weights
                save_model_checkpoint(fabric, model, os.path.join(out_dir, f"iter-{iter_num:06d}.pth"))

        dt = time.time() - t0
        if iter_num % log_interval == 0:
            fabric.print(f"iter {iter_num}: loss {loss.item():.4f}, time: {dt*1000:.2f}ms")

**Comment by xunfeng1980** [comment]
@chakt 3q,It work for me.
![Xnip2023-05-03_20-25-03](https://user-images.githubusercontent.com/8380489/235914872-830c4ce7-2edc-4371-a526-38773d203b00.jpg)

**Comment by xunfeng1980** [comment]
```
    for iter_num in range(starting_iter, starting_iter + max_iters):
```
to 
```
for iter_num in range(starting_iter,  max_iters):
```
